openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,602 @@
+        } else if let Some(node) = self.original_tree.node_by_index(node_index) {
+            Some(node)
+            // If it isn't in the tree either, something has gone wrong.
+        } else {
+            None
+        }
+    }
+
+    /// Returns a mutable reference to the node in the diff at index
+    /// `node_index`. If the diff doesn't have a node at that index, it clones
+    /// the node to the diff and returns a mutable reference to that node.
+    /// Returns an error if the node can neither be found in the tree nor in the
+    /// diff, or if the index is out of the bounds of the diff.
+    fn node_mut_by_index(&mut self, node_index: NodeIndex) -> Result<&mut T, ABinaryTreeDiffError> {
+        // We first check if the given node_index is within the bounds of the diff.
+        if node_index >= self.size() {
+            Err(ABinaryTreeDiffError::OutOfBounds)
+            // We then check if the node is already in the diff. (Not using `if let
+            // ...` here, because the borrow checker doesn't like that).
+        } else if self.diff.contains_key(&node_index) {
+            self.diff
+                .get_mut(&node_index)
+                // We just checked that this index exists, so this must be Some.
+                .ok_or(ABinaryTreeDiffError::LibraryError)
+            // If not, we take a copy from the original tree and put it in the
+            // diff before returning a mutable reference to it.
+        } else if let Some(tree_node) = self.original_tree.node_by_index(node_index) {
+            self.add_node(node_index, tree_node.clone())?;
+            self.diff
+                .get_mut(&node_index)
+                // We just inserted this into the diff, so this should be Some.
+                .ok_or(ABinaryTreeDiffError::LibraryError)
+        } else {
+            // If the node is neither out of bounds, nor in the diff, nor in the
+            // tree, something must have gone wrong somewhere.
+            Err(ABinaryTreeDiffError::LibraryError)
+        }
+    }
+
+    // Helper functions for node addition and removal
+
+    /// Add a node to the diff at the given index. This function can be used
+    /// both to extend the diff (by choosing a `node_index` that is equal to the
+    /// size of the diff) and to place a node at the given index such that any
+    /// previous node in the tree at the same position is replaced upon merging
+    /// the diff. Returns an error if adding the node would increase the size of
+    /// the diff beyond `NodeIndex::max_value()` or if the given node index is
+    /// larger than the current size of the diff.
+    fn add_node(&mut self, node_index: NodeIndex, node: T) -> Result<(), ABinaryTreeDiffError> {
+        // Prevent the tree from becoming too large.
+        if self.size() == NodeIndex::max_value() {
+            return Err(ABinaryTreeDiffError::TreeTooLarge);
+        }
+        // Check that we're extending the tree by at most one.
```

Comment:
I think doing extending and replacing in the same function is error-prone. How about doing to separate functions?